From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Mon, 16 Dec 2024 07:33:52 +0100
Subject: [PATCH 073/225] usb: typec: tcpm: Add support for SC27xx USB Type-C
 PD controller
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Add support for the USB power delivery and Type-C controller found in
newer SC27xx PMICs starting with SC2730. This driver is based on the
original driver [1] from Unisoc's downstream kernel, but has been
cleaned up and simplified a lot.

Although the PMIC actually uses two separate sets of registers for
Type-C and PD, with older PMICs only supporting the Type-C registers,
controlling both in the same driver avoids complicated links between
drivers. The separate Type-C driver [2] for SC2730 is no longer needed,
but older PMICs without PD support may still require a separate driver.

[1]: https://github.com/MotorolaMobilityLLC/kernel-sprd/blob/android-14-release-ula34/drivers/usb/typec/tcpm/sc27xx_pd.c
[2]: https://lore.kernel.org/all/20231211074120.27958-1-xinhu.wu@unisoc.com/

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 drivers/usb/typec/tcpm/Kconfig     |    7 +
 drivers/usb/typec/tcpm/Makefile    |    1 +
 drivers/usb/typec/tcpm/sc27xx_pd.c | 1745 ++++++++++++++++++++++++++++
 3 files changed, 1753 insertions(+)
 create mode 100644 drivers/usb/typec/tcpm/sc27xx_pd.c

diff --git a/drivers/usb/typec/tcpm/Kconfig b/drivers/usb/typec/tcpm/Kconfig
index 00baa7503..fd39e1c9d 100644
--- a/drivers/usb/typec/tcpm/Kconfig
+++ b/drivers/usb/typec/tcpm/Kconfig
@@ -91,4 +91,11 @@ config TYPEC_QCOM_PMIC
 	  This driver enables Type-C role switching, orientation, Alternate
 	  mode and Power Delivery support both for VBUS and VCONN.
 
+config TYPEC_SC27XX_PD
+	tristate "Unisoc SC27xx series PMIC USB PD driver"
+	depends on MFD_SC27XX_PMIC
+	help
+	  This driver adds support for USB PD and Type-C
+	  functionalities sprovided by Unisoc SC27xx PMICs.
+
 endif # TYPEC_TCPM
diff --git a/drivers/usb/typec/tcpm/Makefile b/drivers/usb/typec/tcpm/Makefile
index 7a8cad0c0..4ed21dc0b 100644
--- a/drivers/usb/typec/tcpm/Makefile
+++ b/drivers/usb/typec/tcpm/Makefile
@@ -10,3 +10,4 @@ obj-$(CONFIG_TYPEC_TCPCI_MT6370)	+= tcpci_mt6370.o
 obj-$(CONFIG_TYPEC_TCPCI_MAXIM)		+= tcpci_maxim.o
 tcpci_maxim-y				+= tcpci_maxim_core.o maxim_contaminant.o
 obj-$(CONFIG_TYPEC_QCOM_PMIC)		+= qcom/
+obj-$(CONFIG_TYPEC_SC27XX_PD)		+= sc27xx_pd.o
diff --git a/drivers/usb/typec/tcpm/sc27xx_pd.c b/drivers/usb/typec/tcpm/sc27xx_pd.c
new file mode 100644
index 000000000..cba02da28
--- /dev/null
+++ b/drivers/usb/typec/tcpm/sc27xx_pd.c
@@ -0,0 +1,1745 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (C) 2019 Spreadtrum Communications Inc.
+
+#include <linux/delay.h>
+#include <linux/interrupt.h>
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/mutex.h>
+#include <linux/nvmem-consumer.h>
+#include <linux/of.h>
+#include <linux/of_device.h>
+#include <linux/platform_device.h>
+#include <linux/regmap.h>
+#include <linux/regulator/consumer.h>
+#include <linux/usb/typec.h>
+#include <linux/usb/tcpm.h>
+#include <linux/usb/pd.h>
+#include <linux/usb/typec_dp.h>
+
+/* PMIC global registers definition */
+#define SC27XX_MODULE_EN		0x1808
+#define SC27XX_ARM_CLK_EN0		0x180c
+#define SC27XX_RTC_CLK_EN0		0x1810
+#define SC27XX_XTL_WAIT_CTRL0		0x1b78
+#define UMP9620_MODULE_EN		0x2008
+#define UMP9620_ARM_CLK_EN0		0x200c
+#define UMP9620_RTC_CLK_EN0		0x2010
+#define UMP9620_XTL_WAIT_CTRL0		0x2378
+#define SC27XX_TYPEC_PD_EN		BIT(13)
+#define SC27XX_CLK_PD_SEL		BIT(10)
+#define SC27XX_CLK_PD_EN		BIT(9)
+#define SC27XX_XTL_EN			BIT(8)
+
+/* Typec controller registers definition */
+#define SC27XX_TYPEC_EN			0x0
+#define SC27XX_TYPEC_MODE		0x4
+#define SC27XX_TYPEC_PD_CFG		0x8
+#define SC27XX_TYPEC_INT_EN		0x0c
+#define SC27XX_TYPEC_INT_CLR		0x10
+#define SC27XX_TYPEC_INT_RAW		0x14
+#define SC27XX_TYPEC_INT_MASK		0x18
+#define SC27XX_TYPEC_STATUS		0x1c
+#define SC27XX_TYPEC_RTRIM		0x3c
+#define SC27XX_TYPEC_SW_CFG		0x54
+#define SC27XX_TYPEC_WC_REG		0x58
+#define SC27XX_TYPEC_DBG1		0x60
+#define SC27XX_TYPEC_IBIAS		0x70
+
+/* PD controller registers definition */
+#define SC27XX_PD_TX_BUF		0x0
+#define SC27XX_PD_RX_BUF		0x4
+#define SC27XX_PD_HEAD_CFG		0x8
+#define SC27XX_PD_CTRL			0xc
+#define SC27XX_PD_CFG0			0x10
+#define SC27XX_PD_CFG1			0x14
+#define SC27XX_PD_MESG_ID_CFG		0x18
+#define SC27XX_PD_STS0			0x1c
+#define SC27XX_PD_STS1			0x20
+#define SC27XX_INT_STS			0x24
+#define SC27XX_INT_FLG			0x28
+#define SC27XX_INT_CLR			0x2c
+#define SC27XX_INT_EN			0x30
+#define SC27XX_PD_PHY_CFG0		0x50
+#define SC27XX_PD_PHY_CFG1		0x54
+#define SC27XX_PD_PHY_CFG2		0x58
+
+/* Bit definitions for SC27XX_TYPEC_EN register */
+#define SC27XX_TYPEC_TRY_SRC_EN		BIT(7)
+#define SC27XX_TYPEC_TRY_SINK_EN	BIT(8)
+
+/* Bit definitions for SC27XX_TYPEC_MODE register */
+#define SC27XX_TYPEC_MODE_MASK		GENMASK(1, 0)
+#define SC27XX_TYPEC_RP_LEVEL_MASK	GENMASK(3, 2)
+#define SC27XX_TYPEC_RP_LEVEL(x)	(((x) << 2) & GENMASK(3, 2))
+
+/* Bit definitions for SC27XX_TYPEC_PD_CFG register */
+#define SC27XX_TYPEC_PD_SUPPORT		BIT(0)
+#define SC27XX_TYPEC_PD_CONSTRUCT	BIT(6)
+#define SC27XX_TYPEC_PD_NO_CHEK_DETACH	BIT(9)
+#define SC27XX_TYPEC_SW_FORCE_CC(x)	(((x) << 10) & GENMASK(11, 10))
+#define SC27XX_TYPEC_VCCON_LDO_RDY	BIT(12)
+#define SC27XX_TYPEC_VCCON_LDO_EN	BIT(13)
+
+/* Bit definitions for SC27XX_TYPEC_INT_EN register */
+#define SC27XX_TYPEC_ATTACH_INT		BIT(0)
+#define SC27XX_TYPEC_DETACH_INT		BIT(1)
+
+/* Bit definitions for SC27XX_TYPEC_STATUS register */
+#define SC27XX_TYPEC_CURRENT_STATUS	GENMASK(4, 0)
+#define SC27XX_TYPEC_FINAL_SWITCH	BIT(5)
+#define SC27XX_TYPEC_VBUS_CL(x)		(((x) & GENMASK(7, 6)) >> 6)
+
+/* Bit definitions for SC27XX_TYPEC_WC_REG register */
+#define SC27XX_TYPEC_PD_PR_SWAP		BIT(2)
+#define SC27XX_TYPEC_PD_VCONN_SWAP	BIT(1)
+
+/* Bit definitions for SC27XX_TYPEC_DBG1 register */
+#define SC27XX_TYPEC_VBUS_OK		BIT(8)
+#define SC27XX_TYPEC_CONN_CC		BIT(9)
+
+/* Bit definitions for SC27XX_TYPEC_IBIAS register */
+#define SC27XX_TYPEC_RX_REF_DECREASE	BIT(7)
+
+/* Bit definitions for SC27XX_PD_HEAD_CFG register */
+#define SC27XX_PD_EXTHEAD		BIT(15)
+#define SC27XX_PD_NUM_DO(x)		(((x) << 12) & GENMASK(14, 12))
+#define SC27XX_PD_MESS_ID(x)		(((x) << 9) & GENMASK(11, 9))
+#define SC27XX_PD_POWER_ROLE		BIT(8)
+#define SC27XX_PD_SPEC_REV(x)		(((x) << 6) & GENMASK(7, 6))
+#define SC27XX_PD_DATA_ROLE		BIT(5)
+#define SC27XX_PD_MESSAGE_TYPE(x)	((x) & GENMASK(4, 0))
+#define SC27XX_PD_SPEC_MASK		GENMASK(7, 6)
+
+/* Bit definitions for SC27XX_PD_CTRL register */
+#define SC27XX_PD_TX_START		BIT(0)
+#define SC27XX_PD_HARD_RESET		BIT(2)
+#define SC27XX_PD_TX_FLUSH		BIT(3)
+#define SC27XX_PD_RX_FLUSH		BIT(4)
+#define SC27XX_PD_FAST_START		BIT(6)
+#define SC27XX_PD_CABLE_RESET		BIT(7)
+#define SC27XX_PD_RX_ID_CLR		BIT(8)
+#define SC27XX_PD_RP_SINKTXNG_CLR	BIT(9)
+
+/* Bit definitions for SC27XX_PD_CFG0 register */
+#define SC27XX_PD_RP_CONTROL(x)		((x) & GENMASK(1, 0))
+#define SC27XX_PD_SINK_RP		GENMASK(3, 2)
+#define SC27XX_PD_EN_SOP1_TX		BIT(6)
+#define SC27XX_PD_EN_SOP2_TX		BIT(7)
+#define SC27XX_PD_EN_SOP		BIT(8)
+#define SC27XX_PD_SRC_SINK_MODE		BIT(9)
+#define SC27XX_PD_CTL_EN		BIT(10)
+#define SC27XX_PD_BIST_MODE_EN		BIT(11)
+
+/* Bit definitions for SC27XX_PD_CFG1 register */
+#define SC27XX_PD_AUTO_RETRY		BIT(0)
+#define SC27XX_PD_RETRY(x)		(((x) << 1) & GENMASK(2, 1))
+#define SC27XX_PD_HEADER_REG_EN		BIT(3)
+#define SC27XX_PD_EN_SOP1_RX		BIT(6)
+#define SC27XX_PD_EN_SOP2_RX		BIT(7)
+#define SC27XX_PD_EN_SOP_RX		BIT(8)
+#define SC27XX_PD_RX_AUTO_GOOD_CRC	BIT(9)
+#define SC27XX_PD_FRS_DETECT_EN		BIT(10)
+#define SC27XX_PD_PHY_13M		BIT(11)
+#define SC27XX_PD_TX_AUTO_GOOD_CRC	BIT(12)
+
+/* Bit definitions for SC27XX_PD_MESG_ID_CFG register */
+#define SC27XX_PD_MESS_ID_TX(x)		((x) & GENMASK(2, 0))
+#define SC27XX_PD_MESS_ID_RX(x)		(((x) << 4) & GENMASK(6, 4))
+#define SC27XX_PD_RETRY_MASK		GENMASK(2, 1)
+#define SC27XX_PD_MESS_ID_MASK		GENMASK(2, 0)
+
+/* Bit definitions for SC27XX_TYPEC_SW_CFG register */
+#define SC27XX_TYPEC_SW_SWITCH_MASK	GENMASK(11, 10)
+#define SC27XX_TYPEC_SW_SWITCH(x)	(((x) << 10) & GENMASK(11, 10))
+
+/* Bit definitions for SC27XX_INT_FLG register */
+#define SC27XX_PD_HARD_RST_FLAG		BIT(0)
+#define SC27XX_PD_CABLE_RST_FLAG	BIT(1)
+#define SC27XX_PD_SOFT_RST_FLAG		BIT(2)
+#define SC27XX_PD_PS_RDY_FLAG		BIT(3)
+#define SC27XX_PD_PKG_RV_FLAG		BIT(4)
+#define SC27XX_PD_TX_OK_FLAG		BIT(5)
+#define SC27XX_PD_TX_ERROR_FLAG		BIT(6)
+#define SC27XX_PD_TX_COLLSION_FLAG	BIT(7)
+#define SC27XX_PD_PKG_RV_ERROR_FLAG	BIT(8)
+#define SC27XX_PD_FRS_RV_FLAG		BIT(9)
+#define SC27XX_PD_RX_FIFO_OVERFLOW_FLAG	BIT(10)
+
+/* Bit definitions for SC27XX_PD_STS1 register */
+#define SC27XX_PD_RX_DATA_NUM_MASK	GENMASK(8, 0)
+#define SC27XX_PD_RX_EMPTY		BIT(13)
+
+/* Bit definitions for SC27XX_INT_CLR register */
+#define SC27XX_PD_HARD_RST_RV_CLR	BIT(0)
+#define SC27XX_PD_CABLE_RST_RV_CLR	BIT(1)
+#define SC27XX_PD_SOFT_RST_RV_CLR	BIT(2)
+#define SC27XX_PD_PS_RDY_CLR		BIT(3)
+#define SC27XX_PD_PKG_RV_CLR		BIT(4)
+#define SC27XX_PD_TX_OK_CLR		BIT(5)
+#define SC27XX_PD_TX_ERROR_CLR		BIT(6)
+#define SC27XX_PD_TX_COLLSION_CLR	BIT(7)
+#define SC27XX_PD_PKG_RV_ERROR_CLR	BIT(8)
+#define SC27XX_PD_FRS_RV_CLR		BIT(9)
+#define SC27XX_PD_RX_FIFO_OVERFLOW_CLR	BIT(10)
+
+/* Bit definitions for SC27XX_INT_EN register */
+#define SC27XX_PD_HARD_RST_RV_EN	BIT(0)
+#define SC27XX_PD_CABLE_RST_RV_EN	BIT(1)
+#define SC27XX_PD_SOFT_RST_RV_EN	BIT(2)
+#define SC27XX_PD_PS_RDY_EN		BIT(3)
+#define SC27XX_PD_PKG_RV_EN		BIT(4)
+#define SC27XX_PD_TX_OK_EN		BIT(5)
+#define SC27XX_PD_TX_ERROR_EN		BIT(6)
+#define SC27XX_PD_TX_COLLSION_EN	BIT(7)
+#define SC27XX_PD_PKG_RV_ERROR_EN	BIT(8)
+#define SC27XX_PD_FRS_RV_EN		BIT(9)
+#define SC27XX_PD_RX_FIFO_OVERFLOW_EN	BIT(10)
+
+/* SC27XX_PD_PHY_CFG1 */
+#define SC27XX_PD_CC1_SW		BIT(3)
+#define SC27XX_PD_CC2_SW		BIT(2)
+
+/* SC27XX_PD_PHY_CFG2 */
+#define SC27XX_PD_CFG2_PD_CLK_BIT	BIT(6)
+#define SC27XX_PD_CFG2_RX_REF_CAL_BIT	BIT(7)
+#define SC27XX_PD_CFG2_RX_REF_CAL_SHIFT	7
+
+#define SC27XX_TX_RX_BUF_MASK		GENMASK(15, 0)
+
+#define SC27XX_PD_CFG0_RCCAL_MASK	GENMASK(8, 6)
+#define SC27XX_PD_CFG0_VTL_MASK		GENMASK(3, 2)
+#define SC27XX_PD_CFG0_VTH_MASK		GENMASK(1, 0)
+#define SC27XX_PD_CFG0_RCCAL_SHIFT	6
+#define SC27XX_PD_CFG0_VTL_SHIFT	2
+#define SC27XX_PD_CFG1_VREF_SEL_MASK	GENMASK(7, 6)
+#define SC27XX_PD_CFG1_REF_CAL_MASK	GENMASK(14, 12)
+#define SC27XX_PD_CFG1_VREF_SEL_SHIFT	6
+#define SC27XX_PD_CFG1_REF_CAL_SHIFT	12
+
+#define SC27XX_PD_DATA_MASK		GENMASK(15, 0)
+#define SC27XX_INT_CLR_MASK		0x3fff
+#define SC27XX_INT_EN_MASK		0x85ff
+
+/* chunked extended message */
+#define SC27XX_CHUNKED_EXT_MSG_MASK	GENMASK(7, 0)
+#define SC27XX_CHUNKED_EXT_MSG_SHIFT	8
+
+/* Timeout (us) for pd data ready according to pd datasheet */
+#define SC27XX_PD_RDY_TIMEOUT		2000
+#define SC27XX_PD_RDY_POLL_INTERVAL	50
+
+/* Timeout (us) for vbus ready according to need */
+#define SC27XX_VBUS_OK_TIMEOUT		500000
+#define SC27XX_VBUS_OK_POLL_INTERVAL	100
+
+#define PMIC_SC2730			1
+#define PMIC_UMP9620			2
+
+/* values for SC27XX_TYPEC_CURRENT_STATUS field */
+enum sc27xx_state {
+	SC27XX_DETACHED_SNK = 0,
+	SC27XX_ATTACHWAIT_SNK,
+	SC27XX_ATTACHED_SNK,
+	SC27XX_DETACHED_SRC,
+	SC27XX_ATTACHWAIT_SRC,
+	SC27XX_ATTACHED_SRC,
+	SC27XX_POWERED_CABLE,
+	SC27XX_AUDIO_CABLE,
+	SC27XX_DEBUG_CABLE,
+};
+
+/* values for SC27XX_TYPEC_MODE */
+enum sc27xx_typec_mode {
+	SC27XX_MODE_SNK = 0,
+	SC27XX_MODE_SRC,
+	SC27XX_MODE_DRP,
+};
+
+enum sc27xx_pd_clk_mode {
+	SC27XX_PD_CLK_D6M_A6M = 0,
+	SC27XX_PD_CLK_D13M_A6P5M,
+	SC27XX_PD_CLK_D13M_A13M,
+};
+
+struct sc27xx_pd_variant_data {
+	u32 id;
+	u32 module_en;
+	u32 arm_clk_en0;
+	u32 rtc_clk_en0;
+	u32 xtl_wait_ctrl0;
+};
+
+static const struct sc27xx_pd_variant_data sc2730_data = {
+	.id = PMIC_SC2730,
+	.module_en = SC27XX_MODULE_EN,
+	.arm_clk_en0 = SC27XX_ARM_CLK_EN0,
+	.rtc_clk_en0 = SC27XX_RTC_CLK_EN0,
+	.xtl_wait_ctrl0 = SC27XX_XTL_WAIT_CTRL0,
+};
+
+static const struct sc27xx_pd_variant_data ump9620_data = {
+	.id = PMIC_UMP9620,
+	.module_en = UMP9620_MODULE_EN,
+	.arm_clk_en0 = UMP9620_ARM_CLK_EN0,
+	.rtc_clk_en0 = UMP9620_RTC_CLK_EN0,
+	.xtl_wait_ctrl0 = UMP9620_XTL_WAIT_CTRL0,
+};
+
+struct sc27xx_pd {
+	struct device *dev;
+	struct tcpm_port *tcpm_port;
+	struct work_struct typec_detect_work;
+	struct regmap *regmap;
+	struct tcpc_dev tcpc;
+	struct mutex lock;
+	struct regulator *vbus;
+	struct regulator *vconn;
+	const struct sc27xx_pd_variant_data *var_data;
+	enum typec_cc_polarity cc_polarity;
+	enum typec_cc_status cc1;
+	enum typec_cc_status cc2;
+	enum sc27xx_state state;
+	bool detect_pending;
+	bool constructed;
+	bool vbus_present;
+	bool vbus_on;
+	bool vconn_on;
+	u32 base;
+	u32 typec_base;
+	u32 rc_cal;
+	u32 delta_cal;
+	u32 ref_cal;
+	bool typec_online;
+};
+
+static inline struct sc27xx_pd *tcpc_to_sc27xx_pd(struct tcpc_dev *tcpc)
+{
+	return container_of(tcpc, struct sc27xx_pd, tcpc);
+}
+
+static int sc27xx_pd_clk_cfg(struct sc27xx_pd *pd)
+{
+	int ret;
+
+	ret = regmap_set_bits(pd->regmap, pd->var_data->module_en,
+			      SC27XX_TYPEC_PD_EN);
+	if (ret)
+		return ret;
+
+	ret = regmap_set_bits(pd->regmap, pd->var_data->arm_clk_en0,
+			      SC27XX_CLK_PD_EN);
+	if (ret)
+		return ret;
+
+	return regmap_set_bits(pd->regmap, pd->var_data->xtl_wait_ctrl0,
+			       SC27XX_XTL_EN);
+}
+
+static int sc27xx_pd_disable_clk(struct sc27xx_pd *pd)
+{
+	int ret;
+
+	ret = regmap_clear_bits(pd->regmap, pd->var_data->module_en,
+				SC27XX_TYPEC_PD_EN);
+	if (ret)
+		return ret;
+
+	return regmap_clear_bits(pd->regmap, pd->var_data->arm_clk_en0,
+				 SC27XX_CLK_PD_EN);
+}
+
+static int sc27xx_pd_start_drp_toggling(struct tcpc_dev *tcpc,
+					enum typec_port_type port_type,
+					enum typec_cc_status cc)
+{
+	struct sc27xx_pd *pd = tcpc_to_sc27xx_pd(tcpc);
+	enum sc27xx_typec_mode mode;
+	int ret;
+
+	mutex_lock(&pd->lock);
+
+	switch (port_type) {
+	case TYPEC_PORT_SRC:
+		mode = SC27XX_MODE_SRC;
+		break;
+	case TYPEC_PORT_SNK:
+		mode = SC27XX_MODE_SNK;
+		break;
+	case TYPEC_PORT_DRP:
+		mode = SC27XX_MODE_DRP;
+		break;
+	}
+
+	ret = regmap_update_bits(pd->regmap, pd->typec_base + SC27XX_TYPEC_MODE,
+				 SC27XX_TYPEC_MODE_MASK, mode);
+
+	/* make sure already connected cables are detected */
+	if (!pd->detect_pending) {
+		pd->detect_pending = true;
+		schedule_work(&pd->typec_detect_work);
+	}
+
+	mutex_unlock(&pd->lock);
+
+	return ret;
+}
+
+static int sc27xx_pd_set_cc(struct tcpc_dev *tcpc, enum typec_cc_status cc)
+{
+	struct sc27xx_pd *pd = tcpc_to_sc27xx_pd(tcpc);
+	u32 cfg, rp_level;
+	int ret;
+
+	mutex_lock(&pd->lock);
+
+	switch (cc) {
+	case TYPEC_CC_RD:
+		cfg = SC27XX_TYPEC_TRY_SINK_EN;
+		rp_level = 1;
+		break;
+	case TYPEC_CC_RP_DEF:
+		cfg = SC27XX_TYPEC_TRY_SRC_EN;
+		rp_level = 1;
+		break;
+	case TYPEC_CC_RP_1_5:
+		cfg = SC27XX_TYPEC_TRY_SRC_EN;
+		rp_level = 2;
+		break;
+	case TYPEC_CC_RP_3_0:
+		cfg = SC27XX_TYPEC_TRY_SRC_EN;
+		rp_level = 3;
+		break;
+	default:
+		cfg = 0;
+		rp_level = 1;
+		break;
+	}
+
+	ret = regmap_update_bits(pd->regmap, pd->typec_base + SC27XX_TYPEC_MODE,
+				 SC27XX_TYPEC_RP_LEVEL_MASK,
+				 SC27XX_TYPEC_RP_LEVEL(rp_level));
+	if (ret < 0)
+		goto out;
+
+	ret = regmap_update_bits(pd->regmap, pd->typec_base + SC27XX_TYPEC_EN,
+				 SC27XX_TYPEC_TRY_SINK_EN | SC27XX_TYPEC_TRY_SRC_EN,
+				 cfg);
+
+out:
+	mutex_unlock(&pd->lock);
+
+	return ret;
+}
+
+static int sc27xx_pd_get_cc(struct tcpc_dev *tcpc,
+			    enum typec_cc_status *cc1,
+			    enum typec_cc_status *cc2)
+{
+	struct sc27xx_pd *pd = tcpc_to_sc27xx_pd(tcpc);
+
+	mutex_lock(&pd->lock);
+	*cc1 = pd->cc1;
+	*cc2 = pd->cc2;
+	mutex_unlock(&pd->lock);
+
+	return 0;
+}
+
+static int sc27xx_pd_set_vbus(struct tcpc_dev *tcpc, bool on, bool charge)
+{
+	struct sc27xx_pd *pd = tcpc_to_sc27xx_pd(tcpc);
+	int ret = 0;
+
+	mutex_lock(&pd->lock);
+
+	if (!pd->vbus || pd->vbus_on == on)
+		goto out;
+
+	pd->vbus_on = on;
+
+	if (on)
+		ret = regulator_enable(pd->vbus);
+	else
+		ret = regulator_disable(pd->vbus);
+
+	if (ret < 0)
+		goto out;
+
+	/* no VBUS interrupts are sent, so polling is needed */
+	if (!pd->detect_pending) {
+		pd->detect_pending = true;
+		schedule_work(&pd->typec_detect_work);
+	}
+
+out:
+	mutex_unlock(&pd->lock);
+
+	return ret;
+}
+
+static int sc27xx_pd_get_vbus(struct tcpc_dev *tcpc)
+{
+	struct sc27xx_pd *pd = tcpc_to_sc27xx_pd(tcpc);
+
+	return pd->vbus_present;
+}
+
+static int sc27xx_pd_set_current_limit(struct tcpc_dev *dev, u32 max_ma, u32 mv)
+{
+	return 0;
+}
+
+static int sc27xx_pd_get_current_limit(struct tcpc_dev *dev)
+{
+	return 0;
+}
+
+static int sc27xx_pd_set_polarity(struct tcpc_dev *tcpc,
+				  enum typec_cc_polarity polarity)
+{
+	return 0;
+}
+
+static int sc27xx_pd_set_roles(struct tcpc_dev *tcpc, bool attached,
+			       enum typec_role role, enum typec_data_role data)
+{
+	struct sc27xx_pd *pd = tcpc_to_sc27xx_pd(tcpc);
+	int ret;
+
+	mutex_lock(&pd->lock);
+	ret = regmap_assign_bits(pd->regmap,
+				 pd->base + SC27XX_PD_CFG0,
+				 SC27XX_PD_SRC_SINK_MODE,
+				 role != TYPEC_SINK);
+	mutex_unlock(&pd->lock);
+
+	return ret;
+}
+
+static int sc27xx_pd_set_vconn(struct tcpc_dev *tcpc, bool on)
+{
+	struct sc27xx_pd *pd = tcpc_to_sc27xx_pd(tcpc);
+	int ret = 0;
+
+	mutex_lock(&pd->lock);
+
+	if (!pd->vconn || pd->vconn_on == on)
+		goto out;
+
+	pd->vconn_on = on;
+
+	if (on)
+		ret = regulator_enable(pd->vconn);
+	else
+		ret = regulator_disable(pd->vconn);
+
+out:
+	mutex_unlock(&pd->lock);
+
+	return ret;
+}
+
+static int sc27xx_pd_tx_flush(struct sc27xx_pd *pd)
+{
+	return regmap_set_bits(pd->regmap, pd->base + SC27XX_PD_CTRL,
+			       SC27XX_PD_TX_FLUSH);
+}
+
+static int sc27xx_pd_rx_flush(struct sc27xx_pd *pd)
+{
+	return regmap_set_bits(pd->regmap, pd->base + SC27XX_PD_CTRL,
+			       SC27XX_PD_RX_FLUSH);
+}
+
+static int sc27xx_pd_clear_rx_id(struct sc27xx_pd *pd)
+{
+	return regmap_set_bits(pd->regmap, pd->base + SC27XX_PD_CTRL,
+			       SC27XX_PD_RX_ID_CLR);
+}
+
+static int sc27xx_pd_clear_tx_id(struct sc27xx_pd *pd)
+{
+	return regmap_update_bits(pd->regmap, pd->base + SC27XX_PD_MESG_ID_CFG,
+				  SC27XX_PD_MESS_ID_MASK, 0x0);
+}
+
+static int sc27xx_pd_reset(struct sc27xx_pd *pd, bool clear_tx_id)
+{
+	int ret;
+
+	ret = sc27xx_pd_rx_flush(pd);
+	if (ret < 0)
+		return ret;
+
+	ret = sc27xx_pd_tx_flush(pd);
+	if (ret < 0)
+		return ret;
+
+	ret = sc27xx_pd_clear_rx_id(pd);
+	if (ret < 0)
+		return ret;
+
+	if (clear_tx_id) {
+		ret = sc27xx_pd_clear_tx_id(pd);
+		if (ret < 0)
+			return ret;
+	}
+
+	return 0;
+}
+
+static int sc27xx_pd_send_hardreset(struct sc27xx_pd *pd)
+{
+	int ret;
+
+	ret = sc27xx_pd_reset(pd, true);
+	if (ret < 0) {
+		dev_err(pd->dev, "cannot PD reset, ret=%d\n", ret);
+		return ret;
+	}
+
+	ret = regmap_set_bits(pd->regmap, pd->base + SC27XX_PD_CTRL,
+			      SC27XX_PD_HARD_RESET);
+	if (ret < 0)
+		return ret;
+
+	return 0;
+}
+
+static int sc27xx_pd_set_rx(struct tcpc_dev *tcpc, bool on)
+{
+	struct sc27xx_pd *pd = tcpc_to_sc27xx_pd(tcpc);
+	int ret;
+
+	mutex_lock(&pd->lock);
+	ret = sc27xx_pd_reset(pd, false);
+	if (ret < 0)
+		goto done;
+
+	if (on) {
+		ret = regmap_set_bits(pd->regmap, pd->base + SC27XX_INT_EN,
+				      SC27XX_PD_PKG_RV_EN);
+		if (ret < 0)
+			goto done;
+
+		ret = regmap_set_bits(pd->regmap, pd->base + SC27XX_PD_CFG1,
+				      SC27XX_PD_RX_AUTO_GOOD_CRC);
+		if (ret < 0)
+			goto done;
+
+		ret = regmap_set_bits(pd->regmap, pd->base + SC27XX_PD_CFG0,
+				      SC27XX_PD_CTL_EN);
+		if (ret < 0)
+			goto done;
+	} else {
+		ret = regmap_clear_bits(pd->regmap, pd->base + SC27XX_PD_CFG0,
+					SC27XX_PD_CTL_EN);
+		if (ret < 0)
+			goto done;
+
+		ret = regmap_clear_bits(pd->regmap, pd->base + SC27XX_INT_EN,
+					SC27XX_PD_PKG_RV_EN);
+		if (ret < 0)
+			goto done;
+
+		ret = regmap_clear_bits(pd->regmap, pd->base + SC27XX_PD_CFG1,
+					SC27XX_PD_RX_AUTO_GOOD_CRC);
+		if (ret < 0)
+			goto done;
+	}
+
+done:
+	mutex_unlock(&pd->lock);
+	return ret;
+}
+
+static int sc27xx_pd_tx_msg(struct sc27xx_pd *pd, const struct pd_message *msg)
+{
+	u16 header;
+	u32 data_obj_num, data[PD_MAX_PAYLOAD * 2] = {0}, head = 0;
+	int i, ret;
+
+	ret = sc27xx_pd_tx_flush(pd);
+	if (ret < 0)
+		return ret;
+
+	data_obj_num = msg ? pd_header_cnt_le(msg->header) : 0;
+	if (data_obj_num > PD_MAX_PAYLOAD) {
+		dev_err(pd->dev, "pd tmsg too long, num=%d\n", data_obj_num);
+		return -EINVAL;
+	}
+
+	header = msg ? le16_to_cpu(msg->header) : 0;
+	ret = regmap_write(pd->regmap, pd->base + SC27XX_PD_HEAD_CFG, header);
+	if (ret < 0)
+		return ret;
+
+	ret = regmap_read(pd->regmap, pd->base + SC27XX_PD_HEAD_CFG, &head);
+	if (ret < 0)
+		return ret;
+
+	if (msg) {
+		for (i = 0; i < data_obj_num; i++) {
+			data[2 * i] = le32_to_cpu(msg->payload[i]) & SC27XX_PD_DATA_MASK;
+			data[2 * i + 1] = (le32_to_cpu(msg->payload[i]) >> 16) &
+					  SC27XX_PD_DATA_MASK;
+		}
+	}
+
+	for (i = 0; i < data_obj_num * 2; i++) {
+		ret = regmap_write(pd->regmap, pd->base + SC27XX_PD_TX_BUF,
+				   data[i]);
+		if (ret < 0)
+			return ret;
+	}
+
+	return regmap_set_bits(pd->regmap, pd->base + SC27XX_PD_CTRL,
+			       SC27XX_PD_TX_START);
+}
+
+static int sc27xx_pd_transmit(struct tcpc_dev *tcpc,
+			      enum tcpm_transmit_type type,
+			      const struct pd_message *msg,
+			      unsigned int negotiated_rev)
+{
+	struct sc27xx_pd *pd = tcpc_to_sc27xx_pd(tcpc);
+	int ret;
+
+	mutex_lock(&pd->lock);
+	switch (type) {
+	case TCPC_TX_SOP:
+		ret = sc27xx_pd_tx_msg(pd, msg);
+		if (ret < 0)
+			dev_err(pd->dev, "cannot send PD message, ret=%d\n",
+				ret);
+		break;
+	case TCPC_TX_HARD_RESET:
+		ret = sc27xx_pd_send_hardreset(pd);
+		if (ret < 0)
+			dev_err(pd->dev, "cann't send hardreset ret=%d\n", ret);
+		break;
+	default:
+		dev_err(pd->dev, "type %d not supported", type);
+		ret = -EINVAL;
+	}
+	mutex_unlock(&pd->lock);
+
+	return ret;
+}
+
+static int sc27xx_pd_read_ext_message(struct sc27xx_pd *pd, struct pd_message *msg)
+{
+	int ret, i, ext_msg_data_size;
+	u32 data[PD_MAX_PAYLOAD * 2] = {0};
+	u32 ext_msg_header = 0;
+
+	ret = regmap_read(pd->regmap, pd->base + SC27XX_PD_RX_BUF, &ext_msg_header);
+	if (ret < 0)
+		return ret;
+
+	msg->ext_msg.header = cpu_to_le16(ext_msg_header);
+	if (!(le16_to_cpu(msg->ext_msg.header) & PD_EXT_HDR_CHUNKED)) {
+		dev_info(pd->dev, "unchunked ext_msg unsupported\n");
+		goto done;
+	}
+
+	ext_msg_data_size = pd_ext_header_data_size_le(msg->ext_msg.header);
+	if (ext_msg_data_size > PD_EXT_MAX_CHUNK_DATA) {
+		dev_err(pd->dev, "chunked ext_msg too long, data_size=%d\n",
+			ext_msg_data_size);
+		goto done;
+	}
+
+	/*
+	 * According to the datasheet, sc27xx_pd_rx_buf is 16bit,
+	 * but PD protocol source code msg->ext_msg.data is 8bit.
+	 */
+	for (i = 0; i < (ext_msg_data_size + 1) / 2; i++) {
+		ret = regmap_read(pd->regmap, pd->base + SC27XX_PD_RX_BUF,
+				  (u32 *)&data[i]);
+		if (ret < 0)
+			return ret;
+
+		msg->ext_msg.data[2 * i] = data[i] & SC27XX_CHUNKED_EXT_MSG_MASK;
+		if (2 * i + 1 < ext_msg_data_size)
+			msg->ext_msg.data[2 * i + 1] = (data[i] >> SC27XX_CHUNKED_EXT_MSG_SHIFT) &
+						       SC27XX_CHUNKED_EXT_MSG_MASK;
+	}
+
+done:
+	return sc27xx_pd_rx_flush(pd);
+}
+
+static int sc27xx_pd_read_msg_pdo(struct sc27xx_pd *pd, struct pd_message *msg)
+{
+	u32 data_obj_num, data[PD_MAX_PAYLOAD * 2] = {0};
+	int i, ret = 0;
+
+	if (le16_to_cpu(msg->header) & PD_HEADER_EXT_HDR)
+		return 0;
+
+	data_obj_num = pd_header_cnt_le(msg->header);
+	if (data_obj_num > PD_MAX_PAYLOAD) {
+		dev_err(pd->dev, "pd msg too long, num=%d\n", data_obj_num);
+		return -EINVAL;
+	}
+
+	for (i = 0; i < data_obj_num * 2; i++) {
+		ret = regmap_read(pd->regmap, pd->base + SC27XX_PD_RX_BUF, (u32 *)&data[i]);
+		if (ret < 0)
+			return ret;
+	}
+
+	/*
+	 * According to the datasheet, sc27xx_pd_rx_buf is 16bit,
+	 * but PD protocol source code msg->payload is 32bit,
+	 * so need two 16bit assignment one 32bit.
+	 */
+	for (i = 0; i < data_obj_num; i++)
+		msg->payload[i] = cpu_to_le32(data[2 * i + 1] << 16 | data[2 * i]);
+
+	return 0;
+}
+
+static int sc27xx_pd_retry_cnt_cfg(struct sc27xx_pd *pd, struct pd_message *msg)
+{
+	u32 spec = 0;
+	int ret = 0;
+
+	spec = pd_header_rev_le(msg->header);
+	if (spec == 1)
+		ret = regmap_update_bits(pd->regmap, pd->base + SC27XX_PD_CFG1,
+					 SC27XX_PD_RETRY_MASK,
+					 SC27XX_PD_RETRY(3));
+	else if (spec == 2)
+		ret = regmap_update_bits(pd->regmap, pd->base + SC27XX_PD_CFG1,
+					 SC27XX_PD_RETRY_MASK,
+					 SC27XX_PD_RETRY(1));
+
+	return ret;
+}
+
+static int sc27xx_pd_read_message(struct sc27xx_pd *pd, struct pd_message *msg)
+{
+	int ret;
+	u32 header = 0;
+
+	ret = regmap_read(pd->regmap, pd->base + SC27XX_PD_RX_BUF, &header);
+	if (ret < 0)
+		return ret;
+
+	header &= SC27XX_TX_RX_BUF_MASK;
+	msg->header = cpu_to_le16(header);
+
+	ret = sc27xx_pd_retry_cnt_cfg(pd, msg);
+	if (ret)
+		return ret;
+
+	if (le16_to_cpu(msg->header) & PD_HEADER_EXT_HDR) {
+		ret = sc27xx_pd_read_ext_message(pd, msg);
+		if (ret < 0) {
+			dev_err(pd->dev, "failed to read pd ext_msg, ret=%d\n", ret);
+			return ret;
+		}
+
+		tcpm_pd_receive(pd->tcpm_port, msg, TCPC_TX_SOP);
+		return 0;
+	}
+
+	ret = sc27xx_pd_read_msg_pdo(pd, msg);
+	if (ret)
+		return ret;
+
+	if (!pd_header_cnt_le(msg->header) &&
+	    pd_header_type_le(msg->header) == PD_CTRL_GOOD_CRC) {
+		if (!pd->constructed) {
+			ret = regmap_set_bits(pd->regmap,
+					      pd->typec_base + SC27XX_TYPEC_PD_CFG,
+					      SC27XX_TYPEC_PD_CONSTRUCT);
+			if (ret < 0)
+				return ret;
+			pd->constructed = true;
+		}
+		tcpm_pd_transmit_complete(pd->tcpm_port, TCPC_TX_SUCCESS);
+	} else {
+		tcpm_pd_receive(pd->tcpm_port, msg, TCPC_TX_SOP);
+	}
+
+	return sc27xx_pd_rx_flush(pd);
+}
+
+static int sc27xx_pd_rc_ref_cal(struct sc27xx_pd *pd)
+{
+	u32 vol = pd->rc_cal;
+	u32 pd_ref = pd->ref_cal;
+	u32 val = 0;
+	u32 cfg2_bit7, cfg0_vth = 0, cfg0_vtl = 0, typec_ibis = 0;
+	u32 cfg0, cfg2, cfg0_mask, cfg2_mask = 0;
+	int ret;
+
+	/*
+	 * According to the datasheet, depending on the calibration
+	 * voltage, different register values should be configured.
+	 */
+
+	if (vol >= 0 && vol <= 26)
+		val = 0x0;
+	if (vol >= 27 && vol <= 37)
+		val = 0x1;
+	if (vol >= 38 && vol <= 48)
+		val = 0x2;
+	if (vol >= 49 && vol <= 59)
+		val = 0x3;
+	if (vol >= 60 && vol <= 69)
+		val = 0x4;
+	if (vol >= 70 && vol <= 80)
+		val = 0x5;
+	if (vol >= 81 && vol <= 90)
+		val = 0x6;
+	if (vol >= 91 && vol <= 127)
+		val = 0x7;
+
+	switch (pd_ref) {
+	case 0:
+		cfg2_bit7 = 0;
+		cfg0_vth = 0;
+		cfg0_vtl = 2;
+		typec_ibis = 0;
+		break;
+	case 1:
+		cfg2_bit7 = 0;
+		cfg0_vth = 1;
+		cfg0_vtl = 1;
+		typec_ibis = 1;
+		break;
+	case 2:
+		cfg2_bit7 = 0;
+		cfg0_vth = 1;
+		cfg0_vtl = 1;
+		typec_ibis = 0;
+		break;
+	case 3:
+		cfg2_bit7 = 1;
+		cfg0_vth = 1;
+		cfg0_vtl = 1;
+		typec_ibis = 0;
+		break;
+	case 4:
+		cfg2_bit7 = 0;
+		cfg0_vth = 2;
+		cfg0_vtl = 0;
+		typec_ibis = 0;
+		break;
+	}
+
+	ret = regmap_assign_bits(pd->regmap,
+				 pd->typec_base + SC27XX_TYPEC_IBIAS,
+				 SC27XX_TYPEC_RX_REF_DECREASE,
+				 typec_ibis == 1);
+
+	cfg0 = (cfg0_vtl << SC27XX_PD_CFG0_VTL_SHIFT) & SC27XX_PD_CFG0_VTL_MASK;
+	cfg0 |= (val << SC27XX_PD_CFG0_RCCAL_SHIFT) & SC27XX_PD_CFG0_RCCAL_MASK;
+	cfg0 |= cfg0_vth & SC27XX_PD_CFG0_VTH_MASK;
+	cfg0_mask = SC27XX_PD_CFG0_RCCAL_MASK |
+		    SC27XX_PD_CFG0_VTL_MASK | SC27XX_PD_CFG0_VTH_MASK;
+
+	ret = regmap_update_bits(pd->regmap,
+				 pd->base + SC27XX_PD_PHY_CFG0,
+				 cfg0_mask, cfg0);
+	if (ret < 0)
+		return ret;
+
+	cfg2 = (cfg2_bit7 << SC27XX_PD_CFG2_RX_REF_CAL_SHIFT) & 0x80;
+	cfg2_mask = SC27XX_PD_CFG2_RX_REF_CAL_BIT;
+
+	return regmap_update_bits(pd->regmap, pd->base + SC27XX_PD_PHY_CFG2,
+				  cfg2_mask, cfg2);
+}
+
+static int sc27xx_pd_select_clk_mode(struct sc27xx_pd *pd, u8 pd_clk_mode)
+{
+	int ret;
+	u32 arm_clk_en0_mask, pd_cfg1_mask, pd_phy_cfg2_mask;
+
+	switch (pd_clk_mode) {
+	case SC27XX_PD_CLK_D6M_A6M:
+		arm_clk_en0_mask = SC27XX_CLK_PD_SEL;
+		pd_cfg1_mask = 0;
+		pd_phy_cfg2_mask = 0;
+		break;
+	case SC27XX_PD_CLK_D13M_A13M:
+		arm_clk_en0_mask = 0;
+		pd_cfg1_mask = SC27XX_PD_PHY_13M;
+		pd_phy_cfg2_mask = 0;
+		break;
+	case SC27XX_PD_CLK_D13M_A6P5M:
+	default:
+		arm_clk_en0_mask = 0;
+		pd_cfg1_mask = SC27XX_PD_PHY_13M;
+		pd_phy_cfg2_mask = SC27XX_PD_CFG2_PD_CLK_BIT;
+		break;
+	}
+
+	ret = regmap_update_bits(pd->regmap, pd->var_data->arm_clk_en0,
+				 SC27XX_CLK_PD_SEL, arm_clk_en0_mask);
+	if (ret < 0)
+		return ret;
+
+	ret = regmap_update_bits(pd->regmap, pd->base + SC27XX_PD_CFG1,
+				 SC27XX_PD_PHY_13M, pd_cfg1_mask);
+	if (ret < 0)
+		return ret;
+
+	ret = regmap_update_bits(pd->regmap, pd->base + SC27XX_PD_PHY_CFG2,
+				 SC27XX_PD_CFG2_PD_CLK_BIT, pd_phy_cfg2_mask);
+	if (ret < 0)
+		return ret;
+
+	return 0;
+}
+
+static int sc27xx_pd_get_clk_mode(struct sc27xx_pd *pd)
+{
+	int ret;
+	u32 val;
+
+	ret = regmap_read(pd->regmap, pd->var_data->arm_clk_en0, &val);
+	if (ret < 0)
+		return ret;
+
+	ret = regmap_read(pd->regmap, pd->base + SC27XX_PD_CFG1, &val);
+	if (ret < 0)
+		return ret;
+
+	return regmap_read(pd->regmap, pd->base + SC27XX_PD_PHY_CFG2, &val);
+}
+
+static int sc27xx_pd_delta_cal(struct sc27xx_pd *pd)
+{
+	u32 delta = pd->delta_cal;
+	u32 vol, vref_sel = 0, ref_cal = 0;
+	u32 cfg1, cfg1_mask;
+
+	cfg1_mask = SC27XX_PD_CC1_SW | SC27XX_PD_CC2_SW |
+		    SC27XX_PD_CFG1_VREF_SEL_MASK |
+		    SC27XX_PD_CFG1_REF_CAL_MASK;
+	/*
+	 * According to the datasheet, delta is efuse caliration
+	 * vol = delta * 2 + 1000
+	 */
+	vol = delta * 2 + 1000;
+
+	/*
+	 * According to the datasheet, depending on the calibration
+	 * voltage, different register values should be configured.
+	 */
+
+	if (vol >= 1185 && vol <= 1195) {
+		vref_sel = 0x0;
+		ref_cal = 0x0;
+	} else if (vol >= 1175 && vol < 1185) {
+		vref_sel = 0x0;
+		ref_cal = 0x1;
+	} else if (vol >= 1165 && vol < 1175) {
+		vref_sel = 0x0;
+		ref_cal = 0x2;
+	} else if (vol >= 1155 && vol < 1165) {
+		vref_sel = 0x0;
+		ref_cal = 0x3;
+	} else if (vol >= 1145 && vol < 1155) {
+		vref_sel = 0x0;
+		ref_cal = 0x4;
+	} else if (vol >= 1135 && vol < 1145) {
+		vref_sel = 0x0;
+		ref_cal = 0x5;
+	} else if (vol >= 1125 && vol < 1135) {
+		vref_sel = 0x0;
+		ref_cal = 0x6;
+	} else if (vol >= 1115 && vol < 1125) {
+		vref_sel = 0x1;
+		ref_cal = 0x2;
+	} else if (vol >= 1105 && vol < 1115) {
+		vref_sel = 0x1;
+		ref_cal = 0x3;
+	} else if (vol >= 1095 && vol < 1105) {
+		vref_sel = 0x1;
+		ref_cal = 0x4;
+	} else if (vol >= 1085 && vol < 1095) {
+		vref_sel = 0x1;
+		ref_cal = 0x5;
+	} else if (vol >= 1075 && vol < 1085) {
+		vref_sel = 0x1;
+		ref_cal = 0x6;
+	} else if (vol >= 1065 && vol < 1075) {
+		vref_sel = 0x2;
+		ref_cal = 0x2;
+	} else if (vol >= 1055 && vol < 1065) {
+		vref_sel = 0x2;
+		ref_cal = 0x3;
+	} else if (vol >= 1045 && vol < 1055) {
+		vref_sel = 0x2;
+		ref_cal = 0x4;
+	} else if (vol >= 1035 && vol < 1045) {
+		vref_sel = 0x2;
+		ref_cal = 0x5;
+	} else if (vol >= 1025 && vol < 1035) {
+		vref_sel = 0x2;
+		ref_cal = 0x6;
+	} else if (vol >= 1015 && vol < 1025) {
+		vref_sel = 0x3;
+		ref_cal = 0x2;
+	} else if (vol >= 1005 && vol < 1015) {
+		vref_sel = 0x3;
+		ref_cal = 0x3;
+	} else if (vol >= 1000 && vol < 1005) {
+		vref_sel = 0x3;
+		ref_cal = 0x4;
+	}
+
+	cfg1 = SC27XX_PD_CC1_SW | SC27XX_PD_CC2_SW;
+	cfg1 |= (vref_sel << SC27XX_PD_CFG1_VREF_SEL_SHIFT)
+		& SC27XX_PD_CFG1_VREF_SEL_MASK;
+	cfg1 |= (ref_cal << SC27XX_PD_CFG1_REF_CAL_SHIFT)
+		& SC27XX_PD_CFG1_REF_CAL_MASK;
+
+	return regmap_update_bits(pd->regmap, pd->base + SC27XX_PD_PHY_CFG1,
+				  cfg1_mask, cfg1);
+}
+
+static int sc27xx_pd_module_init(struct sc27xx_pd *pd)
+{
+	int ret;
+
+	ret = sc27xx_pd_delta_cal(pd);
+	if (ret < 0)
+		return ret;
+
+	ret = sc27xx_pd_rc_ref_cal(pd);
+	if (ret < 0)
+		return ret;
+
+	ret = sc27xx_pd_select_clk_mode(pd, SC27XX_PD_CLK_D6M_A6M);
+	if (ret < 0)
+		return ret;
+
+	ret = regmap_update_bits(pd->regmap, pd->base + SC27XX_PD_CFG1,
+				 SC27XX_PD_RETRY_MASK, SC27XX_PD_RETRY(3));
+	if (ret < 0)
+		return ret;
+
+	ret = regmap_set_bits(pd->regmap, pd->typec_base + SC27XX_TYPEC_PD_CFG,
+			      SC27XX_TYPEC_PD_SUPPORT);
+	if (ret < 0)
+		return ret;
+
+	ret = regmap_update_bits(pd->regmap, pd->base + SC27XX_PD_MESG_ID_CFG,
+				 SC27XX_PD_MESS_ID_MASK, 0x0);
+	if (ret < 0)
+		return ret;
+
+	ret = regmap_set_bits(pd->regmap, pd->base + SC27XX_PD_CFG0,
+			      SC27XX_PD_BIST_MODE_EN);
+	if (ret < 0)
+		return ret;
+
+	ret = regmap_write(pd->regmap, pd->base + SC27XX_INT_CLR,
+			   SC27XX_INT_CLR_MASK);
+	if (ret < 0)
+		return ret;
+
+	ret = regmap_clear_bits(pd->regmap, pd->base + SC27XX_PD_CFG1,
+				SC27XX_PD_RX_AUTO_GOOD_CRC);
+	if (ret < 0)
+		return ret;
+
+	sc27xx_pd_get_clk_mode(pd);
+
+	return regmap_write(pd->regmap, pd->base + SC27XX_INT_EN,
+			    SC27XX_INT_EN_MASK);
+}
+
+static int sc27xx_pd_init(struct tcpc_dev *tcpc)
+{
+	struct sc27xx_pd *pd = tcpc_to_sc27xx_pd(tcpc);
+	int ret;
+
+	ret = sc27xx_pd_clk_cfg(pd);
+	if (ret)
+		return ret;
+
+	return sc27xx_pd_module_init(pd);
+}
+
+static irqreturn_t sc27xx_pd_irq(int irq, void *dev_id)
+{
+	struct sc27xx_pd *pd = dev_id;
+	struct pd_message pd_msg;
+	u32 int_sts = 0, pd_sts1 = 0;
+	int ret;
+
+	mutex_lock(&pd->lock);
+	ret = regmap_read(pd->regmap, pd->base + SC27XX_INT_FLG, &int_sts);
+	if (ret < 0)
+		goto done;
+
+	int_sts &= SC27XX_INT_CLR_MASK;
+	ret = regmap_update_bits(pd->regmap, pd->base + SC27XX_INT_CLR,
+				 SC27XX_INT_CLR_MASK, int_sts);
+	if (ret < 0)
+		goto done;
+
+	if (int_sts & SC27XX_PD_HARD_RST_FLAG) {
+		dev_dbg(pd->dev, "received hard reset\n");
+
+		ret = sc27xx_pd_reset(pd, true);
+		if (ret < 0) {
+			dev_err(pd->dev, "cannot reset, ret=%d\n", ret);
+			goto done;
+		}
+
+		tcpm_pd_hard_reset(pd->tcpm_port);
+	}
+
+	if (int_sts & SC27XX_PD_CABLE_RST_FLAG)
+		dev_warn(pd->dev, "received cable reset\n");
+
+	if (int_sts & SC27XX_PD_SOFT_RST_FLAG) {
+		ret = sc27xx_pd_reset(pd, true);
+		if (ret < 0) {
+			dev_err(pd->dev, "cannot reset, ret=%d\n", ret);
+			goto done;
+		}
+	}
+
+	if ((int_sts & SC27XX_PD_PKG_RV_FLAG)) {
+		/*
+		 * According to the requirements of SC2730 ASIC spec, after receiving
+		 * the interrupt,the data should be read after 500us.
+		 */
+		if (pd->var_data->id == PMIC_SC2730) {
+			usleep_range(500, 510);
+		}
+		ret = regmap_read_poll_timeout(pd->regmap,
+					       pd->base + SC27XX_PD_STS1,
+					       pd_sts1,
+					       (pd_sts1 & (~SC27XX_PD_RX_EMPTY)),
+					       SC27XX_PD_RDY_POLL_INTERVAL,
+					       SC27XX_PD_RDY_TIMEOUT);
+		if (ret < 0) {
+			dev_err(pd->dev, "wait for rx status failed, ret = %d\n", ret);
+			goto done;
+		}
+
+		ret = sc27xx_pd_read_message(pd, &pd_msg);
+		if (ret < 0) {
+			dev_err(pd->dev, "failed to read PD msg, ret=%d\n", ret);
+			goto done;
+		}
+	}
+
+	if (int_sts & SC27XX_PD_TX_OK_FLAG)
+		tcpm_pd_transmit_complete(pd->tcpm_port, TCPC_TX_SUCCESS);
+
+	if (int_sts & SC27XX_PD_TX_ERROR_FLAG) {
+		dev_dbg(pd->dev, "PD tx error\n");
+		tcpm_pd_transmit_complete(pd->tcpm_port, TCPC_TX_FAILED);
+	}
+
+	if (int_sts & SC27XX_PD_TX_COLLSION_FLAG) {
+		dev_err(pd->dev, "PD collision\n");
+		tcpm_pd_transmit_complete(pd->tcpm_port, TCPC_TX_FAILED);
+	}
+
+	if (int_sts & SC27XX_PD_PKG_RV_ERROR_FLAG)
+		dev_dbg(pd->dev, "PD rx error\n");
+
+	if (int_sts & SC27XX_PD_RX_FIFO_OVERFLOW_FLAG)
+		dev_err(pd->dev, "PD rx fifo overflow\n");
+
+done:
+	mutex_unlock(&pd->lock);
+
+	return IRQ_HANDLED;
+}
+
+static void sc27xx_cc_polarity_status(struct sc27xx_pd *pd, u32 status)
+{
+	if (status & SC27XX_TYPEC_FINAL_SWITCH)
+		pd->cc_polarity = TYPEC_POLARITY_CC1;
+	else
+		pd->cc_polarity = TYPEC_POLARITY_CC2;
+}
+
+static void sc27xx_cc_status(struct sc27xx_pd *pd, u32 status)
+{
+	u32 cc_rp, rp_sts = SC27XX_TYPEC_VBUS_CL(status);
+
+	switch (rp_sts) {
+	case 0:
+		cc_rp = TYPEC_CC_RP_DEF;
+		break;
+	case 1:
+		cc_rp = TYPEC_CC_RP_1_5;
+		break;
+	case 2:
+		cc_rp = TYPEC_CC_RP_3_0;
+		break;
+	default:
+		cc_rp = TYPEC_CC_OPEN;
+		break;
+	}
+
+	switch (pd->state) {
+	case SC27XX_ATTACHED_SNK:
+	case SC27XX_DEBUG_CABLE:
+		if (pd->cc_polarity == TYPEC_POLARITY_CC1) {
+			pd->cc1 = cc_rp;
+			pd->cc2 = TYPEC_CC_OPEN;
+		} else {
+			pd->cc1 = TYPEC_CC_OPEN;
+			pd->cc2 = cc_rp;
+		}
+		break;
+
+	case SC27XX_ATTACHED_SRC:
+		if (pd->cc_polarity == TYPEC_POLARITY_CC1) {
+			pd->cc1 = TYPEC_CC_RD;
+			pd->cc2 = TYPEC_CC_OPEN;
+		} else {
+			pd->cc1 = TYPEC_CC_OPEN;
+			pd->cc2 = TYPEC_CC_RD;
+		}
+		break;
+
+	case SC27XX_POWERED_CABLE:
+		if (pd->cc_polarity == TYPEC_POLARITY_CC1) {
+			pd->cc1 = TYPEC_CC_RD;
+			pd->cc2 = TYPEC_CC_RA;
+		} else {
+			pd->cc1 = TYPEC_CC_RA;
+			pd->cc2 = TYPEC_CC_RD;
+		}
+		break;
+	default:
+		pd->cc1 = TYPEC_CC_OPEN;
+		pd->cc2 = TYPEC_CC_OPEN;
+		break;
+	}
+
+	tcpm_cc_change(pd->tcpm_port);
+}
+
+static int sc27xx_pd_update_header(struct sc27xx_pd *pd)
+{
+	int ret;
+
+	if (pd->state == SC27XX_ATTACHED_SNK) {
+		ret = regmap_clear_bits(pd->regmap,
+					pd->base + SC27XX_PD_HEAD_CFG,
+					SC27XX_PD_POWER_ROLE);
+		if (ret < 0)
+			return ret;
+
+		ret = regmap_clear_bits(pd->regmap,
+					pd->base + SC27XX_PD_HEAD_CFG,
+					SC27XX_PD_DATA_ROLE);
+		if (ret < 0)
+			return ret;
+	}
+
+	return 0;
+}
+
+static int sc27xx_get_vbus_status(struct sc27xx_pd *pd)
+{
+	u32 status = 0;
+	bool vbus_present;
+	int ret;
+
+	if (pd->typec_online) {
+		ret = regmap_read_poll_timeout(pd->regmap,
+					       pd->typec_base + SC27XX_TYPEC_DBG1,
+					       status,
+					       status & SC27XX_TYPEC_VBUS_OK,
+					       SC27XX_VBUS_OK_POLL_INTERVAL,
+					       SC27XX_VBUS_OK_TIMEOUT);
+		if (ret < 0) {
+			/* vbus may be off when entering source state */
+			if (pd->state == SC27XX_ATTACHED_SRC && !pd->vbus_on)
+				return 0;
+
+			dev_err(pd->dev, "wait for vbus failed, ret = %d\n", ret);
+			return ret;
+		}
+	} else {
+		ret = regmap_read_poll_timeout(pd->regmap,
+					       pd->typec_base + SC27XX_TYPEC_DBG1,
+					       status,
+					       !(status & SC27XX_TYPEC_VBUS_OK),
+					       SC27XX_VBUS_OK_POLL_INTERVAL,
+					       SC27XX_VBUS_OK_TIMEOUT);
+		if (ret < 0) {
+			/* vbus may be on when leaving source state */
+			if (pd->vbus_on)
+				return 0;
+
+			dev_err(pd->dev, "wait for vbus off failed, ret = %d\n", ret);
+			return ret;
+		}
+	}
+
+	vbus_present = !!(status & SC27XX_TYPEC_VBUS_OK);
+	if (vbus_present != pd->vbus_present) {
+		pd->vbus_present = vbus_present;
+		tcpm_vbus_change(pd->tcpm_port);
+	}
+
+	return 0;
+}
+
+static int sc27xx_pd_check_vbus_cc_status(struct sc27xx_pd *pd)
+{
+	u32 val = 0;
+	int ret;
+
+	ret = sc27xx_pd_clk_cfg(pd);
+	if (ret)
+		return ret;
+
+	ret = regmap_read(pd->regmap, pd->typec_base + SC27XX_TYPEC_STATUS,
+			  &val);
+	if (ret)
+		return ret;
+
+	pd->state = val & SC27XX_TYPEC_CURRENT_STATUS;
+
+	if (pd->state == SC27XX_ATTACHED_SNK || pd->state == SC27XX_ATTACHED_SRC ||
+	    pd->state == SC27XX_DEBUG_CABLE)
+		pd->typec_online = true;
+	else
+		pd->typec_online = false;
+
+	sc27xx_pd_update_header(pd);
+	sc27xx_pd_reset(pd, true);
+	sc27xx_cc_polarity_status(pd, val);
+	sc27xx_cc_status(pd, val);
+	sc27xx_get_vbus_status(pd);
+
+	return 0;
+}
+
+static irqreturn_t sc27xx_typec_irq(int irq, void *dev_id)
+{
+	struct sc27xx_pd *pd = dev_id;
+	u32 event;
+	int ret;
+
+	mutex_lock(&pd->lock);
+
+	ret = regmap_read(pd->regmap, pd->typec_base + SC27XX_TYPEC_INT_MASK, &event);
+	if (ret < 0)
+		goto out;
+
+	regmap_write(pd->regmap, pd->typec_base + SC27XX_TYPEC_INT_CLR, event);
+
+	ret = sc27xx_pd_check_vbus_cc_status(pd);
+	if (ret < 0) {
+		dev_err(pd->dev, "failed to check status, ret = %d\n", ret);
+		goto out;
+	}
+
+	mutex_unlock(&pd->lock);
+
+out:
+	return IRQ_HANDLED;
+}
+
+static void sc27xx_init_tcpc_dev(struct sc27xx_pd *pd)
+{
+	pd->tcpc.init = sc27xx_pd_init;
+	pd->tcpc.get_vbus = sc27xx_pd_get_vbus;
+	pd->tcpc.get_current_limit = sc27xx_pd_get_current_limit;
+	pd->tcpc.set_cc = sc27xx_pd_set_cc;
+	pd->tcpc.get_cc = sc27xx_pd_get_cc;
+	pd->tcpc.set_polarity = sc27xx_pd_set_polarity;
+	pd->tcpc.set_vconn = sc27xx_pd_set_vconn;
+	pd->tcpc.set_vbus = sc27xx_pd_set_vbus;
+	pd->tcpc.set_current_limit = sc27xx_pd_set_current_limit;
+	pd->tcpc.set_pd_rx = sc27xx_pd_set_rx;
+	pd->tcpc.set_roles = sc27xx_pd_set_roles;
+	pd->tcpc.start_toggling = sc27xx_pd_start_drp_toggling;
+	pd->tcpc.pd_transmit = sc27xx_pd_transmit;
+}
+
+static int sc27xx_pd_efuse_read(struct sc27xx_pd *pd,
+				const char *cell_id, u32 *val)
+{
+	struct nvmem_cell *cell;
+	void *buf;
+	size_t len = 0;
+
+	cell = nvmem_cell_get(pd->dev, cell_id);
+	if (IS_ERR_OR_NULL(cell))
+		return PTR_ERR(cell);
+
+	buf = nvmem_cell_read(cell, &len);
+	nvmem_cell_put(cell);
+
+	if (IS_ERR(buf))
+		return PTR_ERR(buf);
+
+	memcpy(val, buf, min(len, sizeof(u16)));
+
+	kfree(buf);
+	return 0;
+}
+
+static int sc27xx_pd_cal(struct sc27xx_pd *pd)
+{
+	u32 cc1_cal, cc2_cal;
+	int ret;
+
+	ret = sc27xx_pd_efuse_read(pd, "pdrc_calib", &pd->rc_cal);
+	if (ret)
+		return ret;
+
+	ret = sc27xx_pd_efuse_read(pd, "pddelta_calib", &pd->delta_cal);
+	if (ret)
+		return ret;
+
+	ret = sc27xx_pd_efuse_read(pd, "pdref_calib", &pd->ref_cal);
+	if (ret)
+		return ret;
+
+	ret = sc27xx_pd_efuse_read(pd, "typec_cc1_cal", &cc1_cal);
+	if (ret)
+		return ret;
+
+	ret = sc27xx_pd_efuse_read(pd, "typec_cc2_cal", &cc2_cal);
+	if (ret)
+		return ret;
+
+	ret = regmap_write(pd->regmap, pd->typec_base + SC27XX_TYPEC_RTRIM,
+			   cc1_cal | (cc2_cal << 5));
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+static void sc27xx_typec_int_enable(struct sc27xx_pd *pd)
+{
+	regmap_set_bits(pd->regmap, pd->typec_base + SC27XX_TYPEC_INT_EN,
+			SC27XX_TYPEC_ATTACH_INT | SC27XX_TYPEC_DETACH_INT);
+}
+
+static void sc27xx_pd_detect_typec_work(struct work_struct *work)
+{
+	struct sc27xx_pd *pd = container_of(work, struct sc27xx_pd, typec_detect_work);
+	int ret;
+
+	mutex_lock(&pd->lock);
+
+	ret = sc27xx_pd_check_vbus_cc_status(pd);
+	if (ret < 0)
+		dev_err(pd->dev, "failed to check status, ret = %d\n", ret);
+
+	pd->detect_pending = false;
+
+	mutex_unlock(&pd->lock);
+}
+
+static int sc27xx_pd_probe(struct platform_device *pdev)
+{
+	struct sc27xx_pd *pd;
+	const struct sc27xx_pd_variant_data *pdata;
+	int pd_irq, typec_irq, ret;
+
+	pdata = of_device_get_match_data(&pdev->dev);
+	if (!pdata) {
+		dev_err(&pdev->dev, "No matching driver data found\n");
+		return -EINVAL;
+	}
+
+	pd = devm_kzalloc(&pdev->dev, sizeof(*pd), GFP_KERNEL);
+	if (!pd) {
+		dev_err(&pdev->dev, "failed to alloc memory\n");
+		return -ENOMEM;
+	}
+
+	pd->dev = &pdev->dev;
+	pd->regmap = dev_get_regmap(pdev->dev.parent, NULL);
+	if (!pd->regmap) {
+		dev_err(&pdev->dev, "failed to get pd regmap\n");
+		return -ENODEV;
+	}
+
+	ret = of_property_read_u32_index(pdev->dev.of_node, "reg", 0,
+					 &pd->base);
+	if (ret) {
+		dev_err(&pdev->dev, "failed to get pd reg address\n");
+		return ret;
+	}
+
+	ret = of_property_read_u32_index(pdev->dev.of_node, "reg", 1,
+					 &pd->typec_base);
+	if (ret) {
+		dev_err(&pdev->dev, "failed to get typec reg address\n");
+		return ret;
+	}
+
+	pd_irq = platform_get_irq(pdev, 0);
+	if (pd_irq < 0) {
+		dev_err(&pdev->dev, "failed to get pd irq number\n");
+		return pd_irq;
+	}
+
+	typec_irq = platform_get_irq(pdev, 1);
+	if (typec_irq < 0) {
+		dev_err(&pdev->dev, "failed to get typec irq number\n");
+		return typec_irq;
+	}
+
+	mutex_init(&pd->lock);
+	pd->var_data = pdata;
+	pd->vbus_present = false;
+	pd->constructed = false;
+	pd->tcpc.fwnode = device_get_named_child_node(&pdev->dev, "connector");
+
+	ret = sc27xx_pd_cal(pd);
+	if (ret) {
+		dev_err(&pdev->dev, "failed to calibrate with efuse data\n");
+		return ret;
+	}
+	sc27xx_init_tcpc_dev(pd);
+
+	pd->vbus = devm_regulator_get_exclusive(pd->dev, "vbus");
+	if (IS_ERR(pd->vbus)) {
+		ret = PTR_ERR(pd->vbus);
+		if (ret != -ENODEV) {
+			dev_err(&pdev->dev, "failed to get vbus regulator\n");
+			return ret;
+		}
+		pd->vbus = NULL;
+	} else {
+		if (regulator_is_enabled(pd->vbus) > 0)
+			regulator_disable(pd->vbus);
+	}
+
+	pd->vconn = devm_regulator_get_exclusive(pd->dev, "vconn");
+	if (IS_ERR(pd->vconn)) {
+		ret = PTR_ERR(pd->vconn);
+		if (ret != -ENODEV) {
+			dev_err(&pdev->dev, "failed to get vconn regulator\n");
+			return ret;
+		}
+		pd->vconn = NULL;
+	} else {
+		if (regulator_is_enabled(pd->vbus) > 0)
+			regulator_disable(pd->vbus);
+	}
+
+	ret = devm_request_threaded_irq(pd->dev, pd_irq, NULL,
+					sc27xx_pd_irq,
+					IRQF_ONESHOT | IRQF_TRIGGER_LOW,
+					"sc27xx_pd", pd);
+	if (ret < 0) {
+		dev_err(pd->dev, "failed to request irq\n");
+		return ret;
+	}
+
+	ret = devm_request_threaded_irq(pd->dev, typec_irq, NULL,
+					sc27xx_typec_irq,
+					IRQF_ONESHOT | IRQF_TRIGGER_LOW,
+					"sc27xx_typec", pd);
+	if (ret < 0) {
+		dev_err(pd->dev, "failed to request typec irq\n");
+		return ret;
+	}
+
+	platform_set_drvdata(pdev, pd);
+
+	/* typec_detect_work is not initialized yet, so this must be true */
+	pd->detect_pending = true;
+
+	pd->tcpm_port = tcpm_register_port(pd->dev, &pd->tcpc);
+	if (IS_ERR(pd->tcpm_port)) {
+		dev_err(pd->dev, "failed to register tcpm port\n");
+		return PTR_ERR(pd->tcpm_port);
+	}
+
+	INIT_WORK(&pd->typec_detect_work, sc27xx_pd_detect_typec_work);
+	schedule_work(&pd->typec_detect_work);
+
+	sc27xx_typec_int_enable(pd);
+
+	return 0;
+}
+
+static void sc27xx_pd_remove(struct platform_device *pdev)
+{
+	struct sc27xx_pd *pd = platform_get_drvdata(pdev);
+
+	tcpm_unregister_port(pd->tcpm_port);
+}
+
+#ifdef CONFIG_PM_SLEEP
+static int sc27xx_pd_suspend(struct device *dev)
+{
+	struct sc27xx_pd *pd = dev_get_drvdata(dev);
+	int ret;
+
+	if (pd->typec_online)
+		return 0;
+
+	ret = sc27xx_pd_disable_clk(pd);
+	if (ret) {
+		dev_err(pd->dev, "failed to disable pd clock when suspend, ret = %d\n", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int sc27xx_pd_resume(struct device *dev)
+{
+	struct sc27xx_pd *pd = dev_get_drvdata(dev);
+	int ret;
+
+	ret = sc27xx_pd_clk_cfg(pd);
+	if (ret) {
+		dev_err(pd->dev, "failed to enable pd clock when resume, ret = %d\n", ret);
+		return ret;
+	}
+
+	return 0;
+}
+#endif
+
+static const struct dev_pm_ops sc27xx_pd_pm_ops = {
+	SET_SYSTEM_SLEEP_PM_OPS(sc27xx_pd_suspend, sc27xx_pd_resume)
+};
+
+static const struct of_device_id sc27xx_pd_of_match[] = {
+	{ .compatible = "sprd,sc2730-pd", .data = &sc2730_data },
+	{ .compatible = "sprd,ump9620-pd", .data = &ump9620_data },
+	{ }
+};
+MODULE_DEVICE_TABLE(of, sc27xx_pd_of_match);
+
+static struct platform_driver sc27xx_pd_driver = {
+	.probe = sc27xx_pd_probe,
+	.remove = sc27xx_pd_remove,
+	.driver = {
+		.name = "sc27xx-typec-pd",
+		.of_match_table = sc27xx_pd_of_match,
+		.pm = &sc27xx_pd_pm_ops,
+	},
+};
+
+module_platform_driver(sc27xx_pd_driver);
+
+MODULE_AUTHOR("Freeman Liu <freeman.liu@spreadtrum.com>");
+MODULE_DESCRIPTION("Spreadtrum SC27xx Type-C driver");
+MODULE_LICENSE("GPL v2");
